### ISSUE TITLE
separate production vat bootstrap from testing mechanisms

### ISSUE THREAD

**Issue by dckc** [Agoric/Agoric#4165]
## What is the Problem Being Solved?

The initial conditions for all the vats in our system are in [vats/src/bootstrap.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/vats/src/bootstrap.js). Currently, it mixes a bunch of testing / demo mechanisms in a way that makes it difficult to be confident that we aren't vulnerable to excess authority.

## Description of the Design

`bootstrap.js` currently interacts with a large set of features:
 - setting up zoe
 - setting up the vaults and AMM
   - `getBootstrapPayment` (note plans to remove this in Agoric/agoric-sdk#4021)
   - for testing: create "fake" price authorities
     - note plans to use something like coinmarketcap in Agoric/documentation#664
   - for testing: setting up liquidity in the AMM
 - setting up dapp facilities including http API endpoints and plugins
 - provisioning user accounts
   - creating vattp and comms remote / egress
   - creating IBC ports
   - creating the `home` bundle and connecting it to the REPL and http
   - for testing / devnet: minting tokens so that users start with assets to work with

@michaelfig @warner what's missing in that list? Can you think of any that we should get rid of?

My inclination is to start with a simple bootstrap that establishes the connections suitable for production and then layer testing mechanisms on top of that. Perhaps more of the functionality should be installed via Zoe? For example, the LINK and ATOM we use for testing could come from a contract that creates a new zcfMint and uses the mint to add liquidity to the AMM and to provide each new user with some funds to play with.

cc @erights @Chris-Hibbert @rowgraus 

## Security Considerations

Code such as `vat-mints.js`, which mints tokens freely, is useful for demos, but should be clearly segregated from code used in production.

cc @jessysaurusrex 

## Documentation Considerations

The `home` API is somewhat under-documented. (Separate issues: Agoric/agoric-sdk#4347 , Agoric/agoric-sdk#3154).

## Test Plan

Review the current `bootstrap.js` looking for features that we don't currently have automated tests for and if we want to keep them, add tests.

**Comment by Chris-Hibbert** [comment]
`switch (ROLE)`  allows one bootstrap to handle production validators, production clients, and fake chains.

**Comment by dckc** [comment]
The current bootstrap mints 1B RUN out of thin air for a `feePurse`.
Anybody have thoughts on how this should work in production?

https://github.com/Agoric/agoric-sdk/blob/71f81012390d4aab7f76e8135cff58db954972fc/packages/vats/src/bootstrap.js#L81-L86

https://github.com/Agoric/agoric-sdk/blob/71f81012390d4aab7f76e8135cff58db954972fc/packages/zoe/src/zoeService/feeMint.js#L40-L42

**Comment by Chris-Hibbert** [comment]
If we need a subsidy to get the chain off the ground, we should find a way to keep it small/limited.

As I understand current thinking, at the beginning of the live chain, the only source of capital that's available is BLD tokens. Those could be used to get RUN line of credit, which could be used to pay fees in order to get Vaults (née Treasury) and AMM set up.  In order to fund various pools for the AMM, we'd need a source for other tokens. The Vaults really shouldn't be started until their respective AMM pools exist.

The hard part about bootstrapping is that the RUN line of credit can't be started without a little bit of RUN to pay for fees, so I think the start-up subsidy is necessary. Luckily, the `feeMintAccess` token that we need for Vaults and RUN LoC could also enable a one-time faucet. But the faucet can't run without a source of fees, so whatever work-around we'd use to start the faucet could just be used on the RUN LoC instead.

**Comment by dckc** [comment]
@dtribble notes that genesis for the JS phase may inherit RUN minted during the cosmos-only phase. Should one of the owners spend that to bootstrap the JS components?

Or... can we just not charge fees for some stuff?

**Comment by Chris-Hibbert** [comment]
another alternative from this morning's #engineering discussion: 

* a big switch to turn fee charging on and off. The simplest version just allows bootstrap to proceed without fees until sufficient parts of the economy are running, and then turn on fees.

**Comment by dckc** [comment]
idea from chat with @michaelfig : CapTP + spawner + bootstrap namehub

solo: bootstrap object in web.js

**Comment by dckc** [comment]
It looks like `makeFeeCollector(zoe, creatorFacet)` was broken in in Agoric/agoric-sdk#4026.  There doesn't seem to be a test.  The method should map over the array of facets. @Chris-Hibbert says he changed `schedulePayments()` in `distributeFees.js` to map over something, but didn't do the same with `makeFeeCollector()`.

I think I fixed it in 71a8b3f60, but a bunch of other stuff is mixed in there.

**Comment by Chris-Hibbert** [comment]
> I think I fixed it in 71a8b3f, but a bunch of other stuff is mixed in there.

The lines that seem to be addressing this 
``` javascript
    E(vats.distributeFees)
      .buildDistributor(
        [treasuryCreator, ammFacets.creatorFacet].map(cf =>
          E(vats.distributeFees).makeFeeCollector(zoeWUnlimitedPurse, cf),
        ),
        feeCollectorDepositFacet,
        epochTimerService,
        harden(distributorParams),
      )
```
look right to me.

**Comment by dckc** [comment]
@michaelfig I rebased the [bootstrap-core](https://github.com/Agoric/agoric-sdk/tree/bootstrap-core) branch on `master`: 26d1f6e5d

esp [bootstrap-core.js](https://github.com/Agoric/agoric-sdk/blob/bootstrap-core/packages/vats/src/bootstrap-core.js)

**Comment by dckc** [comment]
Idea: mainnet 1 Genesis names privileged modules to install.

**Comment by michaelfig** [comment]
> Idea: mainnet 1 Genesis names privileged modules to install.

I think that the `genesis.json` could have something like:

`s/Power Snippet/Capsule/g`

```js
  "bootstrapCapsules": ["someFunc1", "someFunc2"],
```

where you define the capsules in `bootstrap-core.js` something like:

```js
// Trick to reify capsule async function names as properties we can invoke.
import * as capsules from './capsules.js'

// ... thread through the bootstrapCapsules as part of `vatParameters.argv`, then:

// Run all the bootstrap capsules in parallel.
await Promise.all(
  bootstrapCapsules.map(
    cname => Promise.resolve().then(() => capsules[cname]({ ....powers, capsules })),
  ),
);
```

**Comment by dckc** [comment]
further refinement (41b985079): let `genesis.json` say exactly which powers each capsule gets.

```js
export const bootstrapManifest = harden({
  behaviors: {
    connectVattpWithMailbox: true,
    buildZoe: true,
  },
  endowments: {
    connectVattpWithMailbox: {
      vatPowers: { D: true },
      vats: { vattp: true },
      devices: { mailbox: true },
    },
    buildZoe: {
      vats: { vatAdmin: true },
      devices: { vatAdmin: true },
      workspace: true,
    },
  },
});

export const simBootstrapManifest = harden({
  behaviors: { installSimEgress: true, ...bootstrapManifest.behaviors },
  endowments: {
    installSimEgress: {
      vatParameters: { argv: { hardcodedClientAddresses: true } },
      vats: {
        vattp: true,
        comms: true,
      },
      workspace: true,
    },
    ...bootstrapManifest.endowments,
  },
});
```

```
2022-01-06T06:39:05.642Z start: swingset running
2022-01-06T06:39:05.643Z outbound: invoking deliverator; 1 new messages for sim-chain
2022-01-06T06:39:05.649Z SwingSet: vat: v1: bootstrap: installSimEgress({"vatParameters":{"argv":{"hardcodedClientAddresses":true}},"vats":{"vattp":true,"comms":true},"workspace":true})
2022-01-06T06:39:05.651Z SwingSet: vat: v1: bootstrap: connectVattpWithMailbox({"vatPowers":{"D":true},"vats":{"vattp":true},"devices":{"mailbox":true}})
2022-01-06T06:39:05.655Z SwingSet: vat: v1: bootstrap: buildZoe({"vats":{"vatAdmin":true},"devices":{"vatAdmin":true},"workspace":true})
```

**Comment by warner** [comment]
One tiny note: the need to connect `vats.comms` and `vats.vattp`, plus the need to connect `devices.mailbox` to `vats.vattp`, are common to all swingset installations, rather than being a specific feature of the Agoric chain or ag-solo instances. I haven't been able to think of a clean way to have swingset do this wiring itself (#51), so it remains as a "some assembly required" wart that host apps (chain/ag-solo) must deal with.

I'd suggest encapsulating that part of the initialization into a separate function, to at least make it clear that this is a part that *ought* to be swingset's responsibility, but unfortunately is not yet.

**Comment by dckc** [comment]
see also: authority viz
https://gist.github.com/dckc/b910588219de97b72e507050564b33c5#file-x-svg
https://github.com/Agoric/agoric-sdk/tree/dc-authority-viz 0ce4971

more recently:
https://gist.github.com/dckc/63b0635baf6f5a4a60af1e6ac75c5b40

**Comment by dckc** [comment]
3 chunks recently:

 1. Thu, 27 Jan PM  57a59088b refactor(vats): start vault factory without install-on-chain
 2. Fri, 28 Jan AM: 2e9baf27f feat(vats): support behavior config via boostrap vatParameters
 3. Fri, 28 Jan PM:
     - c6c79468c chore(run-protocol): fix, test AMM bootstrap
     - cd8afa35a test(run-protcol): unit test startVaultFactory

**Comment by dckc** [comment]
### BLD issuer should go in its own vat for scaling

Re the BLD issuer in bootstrap, I had: `// TODO: should this live in another vat???`

In discussion with @dtribble , he pointed out that we expect _lots_ of BLD purses, and purses go in the same vat as the issuer. So yes, it should go in its own vat. Agoric/agoric-sdk#4578

### Upgrade the bootstrap vat

@dtribble notes we should be able to upgrade the bootstrap vat. Agoric/agoric-sdk#4548

**Comment by dckc** [comment]
It looks like we're moving back to using `boot.js` in order to do #5819, so let's be sure to remove `SIM_CHAIN_BOOTSTRAP_MANIFEST` while we're at it. For `decentral-demo-config.js`, we can use a separate bootstrap module, analogous to `boot.js`.

https://github.com/Agoric/agoric-sdk/blob/416b3e73f269e40a07d83be1da951d4875c0ca61/packages/vats/src/core/boot.js#L13

**Comment by abebeos** [comment]
(@dckc most possibly you disagree with me on this, but this is another example for "confusion caused by reopening an issue". This issue here was closed while assigned to a milestone, which is now closed. Preferably, open a new issue, and point back to this (closed) issue.)

(please just give me an ack that you saw this, and I shred the comment. I shred my comments anyways at some point, but can do it faster so folks remains calm)

**Comment by dckc** [comment]
rather than edit description to re-summarize history back to Dec 2021, let's make a new issue: #6687
